File: icpcmb2025_e.cpp
#pragma GCC optimize("Ofast")
#include <bits/stdc++.h>
using namespace std;

int main() {
    int n, m;
    cin >> n >> m;

    vector<pair<int,int>> edges;
    vector<vector<pair<int,int>>> adj(n+1);

    for (int i = 0; i < m; ++i) {
        int u, v;
        cin >> u >> v;
        edges.emplace_back(u, v);
        adj[u].emplace_back(v, i);
        adj[v].emplace_back(u, i);
    }

    const int LIM = 2000000;
    vector<double> visited(m, 0.0);

    std::mt19937 rng((unsigned)chrono::high_resolution_clock::now().time_since_epoch().count());

    for (int it = 0; it < LIM; ++it) {
        int cur = 1;
        while (cur != n) {
            auto &neis = adj[cur];

            int idx = std::uniform_int_distribution<int>(0, (int)neis.size() - 1)(rng);
            int v = neis[idx].first;
            int eid = neis[idx].second;
            visited[eid] += 1.0;
            cur = v;
        }
    }

    vector<pair<int,double>> edges_;
    edges_.reserve(m);
    for (int i = 0; i < m; ++i) edges_.emplace_back(i, visited[i] / LIM);

    sort(edges_.begin(), edges_.end(),
         [](const pair<int,double>& a, const pair<int,double>& b) {
             if (a.second != b.second) return a.second > b.second;
             return a.first < b.first;
         });

    long double ans = 0.0L;
    for (int i = 0; i < m; ++i) {
        int rank = i + 1;
        long double freq = edges_[i].second;
        ans += freq * rank;
    }

    cout.setf(std::ios::fixed); 
    cout << setprecision(3) << (double)ans << "\n";
}

File: icpcmb2025_e.py
import random
import sys
from collections import defaultdict
n, m = map(int, sys.stdin.readline().split())
edges = []
adj = [[] for _ in range(n+1)]
for i in range(m):
    u, v = map(int, sys.stdin.readline().split())
    edges.append((u, v))
    adj[u].append((v, i))
    adj[v].append((u, i))

lim = 200000  
visited = [0.0] * m

for _ in range(lim):
    cur = 1
    while cur != n:
        v, eid = random.choice(adj[cur])
        visited[eid] += 1
        cur = v

cnt = [x / lim for x in visited]

edges_ = sorted(enumerate(cnt), key=lambda x: -x[1])
ans = 0.0
for rank, (eid, freq) in enumerate(edges_, start=1):
    ans += freq * rank

print(f"{ans:.3f}")